### PR TITLE
Feature/cicd

### PR DESCRIPTION
**Pull request by dylanbannon**
We now have automated cluster creation and destruction tests that are triggered by pushes!

This branch isn't ready to merge yet. This PR is just being filed to get feedback from others.

Here are the remaining points of work/discussion:

[] The tests currently run on Github Actions, but our Docker build appears to be controlled by Travis. We should consolidate these tests and the Docker build together onto either Github Actions or Travis. I don't have any preference which.
[] A lot of environmental variables are defined in `.github/workflows/main.yml`, more than we need. A minimal list of necessary environmental variables should be identified.
[] The service account used for deployment currently has the following roles assigned in Google Cloud: Editor, Compute Admin, Kubernetes Engine Admin, Service Account Admin, and Service Account User. A minimal set of these should be identified and documented, perhaps as comments to the `gke/test/create/service-account` endpoint.
[] The cluster name (currently, the `CLOUDSDK_CONTAINER_CLUSTER` and `CLUSTER_NAME` variables) should be randomized, somehow. This might take a little work. Ideally, it'd be randomized similar to the way that cluster names are randomized by default in the main menu.
[] I made copies of all Makefiles I needed to modify (saved as `[Makefile_name].test`) and left the copies in the same directories as the originals. This might not be ideal for organization. Other organizational options include adding the `test/` targets to existing Makefiles, instead of breaking them out into their own files or, if we're still breaking them out, maybe organizing all the `*.test` Makefiles into their own folder structure. We should talk about the best way to organize this.
[] A necessary prerequisite for automation is installing the `kubens` script. It's committed to the git project and just sitting in the `conf` folder right now, but that looks awkward. I need to revise the `test` Makefile target to install `kubens` on the fly before cluster configuration.
[] There's currently no way to test `ELK` deployment, since the `shopt` command is incompatible with the Github Actions environment. If we switch everything over to Travis, this might get resolved. Otherwise, we should think about whether it's worth replacing the `shopt` calls in order to enable this test.
[] FINALLY: Right now, we're just creating and deleting a cluster, but not testing any cluster functionality. This is a great start, but a lot of the core functionality that we've unintentionally broken in the past (and then spent hours or days repairing) is still outside the tests' scope. We need to add tests for cluster functionality once the cluster is up.

### CODE REVIEW

**Comment by willgraf** [comment]
In regards to Travis CI vs GitHub Actions, I would personally rather use Travis CI merely because I have never used GitHub Actions.

Since our test/build services do not need to be served on the same system, let's make an issue to do some due diligence on which platform we should use and not have this block the current PR.

**Comment by willgraf** [comment]
I think that some small changes that would improve our integration testing would be to affirm that all the health checks for our services are up.  This primarily includes the frontend and tf-serving, but could also include the prometheus stack and redis.

Maybe instead of networks, we can use a helm command to check the status of our pods?

**Comment by dylanbannon** [comment]
@willgraf 

For the first point, I'll do my best to consolidate the production and testing cluster deployment/destruction workflows (i.e., Makefile targets and supporting scripts) as much as possible. You're right that they should share as much code as possible; I only have them setup in parallel like they are now for expediency of development.

For the second point, I completely agree. Spinning up clusters at every commit seems a little excessive. I'm going to pursue two avenues here:
First, I'll check out these `gcloud` docs for some ideas: https://cloud.google.com/functions/docs/testing/test-basics.
Second, if necessary, I'll look into having a switch for committers to turn tests on or off, or maybe only enable them when evaluating pull requests.

**Comment by dylanbannon** [comment]
@willgraf 

I've migrated all the tests to Travis CI (fixing #269) and, in the process, changed the conditions on the cluster tests so that they are only triggered in pull requests.

I still need to consolidate the testing and non-testing workflows as much as possible.

**Comment by dylanbannon** [comment]
@willgraf 

Two things to discuss:

1. Is having cluster deployment tests only on pull requests optimal? Do we want them to be more frequent?

2. The `shopt` command has failed in both GitHub Actions and TravisCI, even though they both seem to be using Bash in their testing environments. Is there some way we can replace `shopt`? For the time being, I'll file a bug on this.

**Comment by willgraf** [comment]
> Is having cluster deployment tests only on pull requests optimal? Do we want them to be more frequent?

I think 2 classes of tests (Integration and Unit) would be useful. Ideally, we can have run tests that check the validity of the commands without running them on every commit. On PRs we can run the full suite of tests since they are more time consuming and cost cloud resources.

> The `shopt` command has failed in both GitHub Actions and TravisCI, even though they both seem to be using Bash in their testing environments. Is there some way we can replace `shopt`? For the time being, I'll file a bug on this.

😞 That's a shame.  I am sure there are other ways to check case-sensitivity in bash, though I found `shopt` to be the easiest to implement.

**Comment by dylanbannon** [comment]
> I think 2 classes of tests (Integration and Unit) would be useful.
Great point. So, for now, it seems like we only have a framework for integration tests, ironically, but no individual unit tests. I'll open an issue on the lack of unit tests, but let me know if your understanding is different from mine. Is it fair to say this this is a non-blocking issue for this PR?

>  I am sure there are other ways to check case-sensitivity in bash, though I found shopt to be the easiest to implement.
Yeah, I can only imagine the Bash hieroglyphics that people on Stack Overflow might suggest for solving this problem. Is there any way to remove the need for case-insensitivity? Is that an easier solution? Unfortunately, I think this issue is blocking for this PR, since I had to modify production code to get the tests to pass.

File: .travis.yml [hunk 1]
```diff
@@ -9,8 +9,51 @@ language: bash
...
+        - GPU_MAX_TIMES_TWO_HUNDRED=800 
+        - GPU_NODE_MIN_SIZE=0
+        - GPU_NODE_MAX_SIZE=4
+        - GPU_PER_NODE=1
+        - KUBERNETES_VERSION=latest
+        - NODE_MIN_SIZE=1
+        - NODE_MAX_SIZE=60
+        - REGION_ZONES_WITH_GPUS=us-west1-a,us-west1-b
+      before_install:
+        - openssl aes-256-cbc -K $encrypted_84460f9bb44c_key -iv $encrypted_84460f9bb44c_iv -in secrets/gcloud_key_base64.enc -out gcloud_key_base64 -d
```
Comment (willgraf) [inline .travis.yml]: Can we change the path from `secrects/gcloud_key.enc` to `.secrets/gcloud_key.enc`? Hiding this file as best as we can feels most user-friendly.

File: conf/tasks/Makefile.gke [hunk 2]
```diff
@@ -234,7 +234,7 @@ gke/destroy/bucket:
 ## Deploy helm tiller with service account
 gke/deploy/helm:
 	@echo "Deploying kubernetes resources..."
-	@kubectl create serviceaccount --namespace kube-system tiller 2> /dev/null || echo "Kubernetes already appears to be deployed."
```
Comment (willgraf) [inline conf/tasks/Makefile.gke]: Ignoring the error output seems like a generally good thing. Should we even have the `|| echo` command? Maybe we should just fail here.

Come to think of it, could be nice to have a way to clean-up failures at this point.

File: conf/tasks/Makefile.gke [hunk 3]
```diff
@@ -275,3 +279,31 @@ gke/destroy/all: \
...
+	@echo "GKE cluster created"
+	@exit 0
+
+## Destroy Cluster
+gke/test/destroy/all: \
+	gke/destroy/node-pools \
+	gke/destroy/cluster \
+	gke/destroy/pds
+	@echo "GKE cluster destroyed"
+	@exit 0
```
Comment (willgraf) [inline conf/tasks/Makefile.gke]: If this fails, how can we clean up the resources that didn't get destroyed?

Comment (dylanbannon) [reply conf/tasks/Makefile.gke]: Great point. I opened #274 off of this comment. This seems to be a problem in production, too.

File: conf/tasks/Makefile.gke (hunk 2 above)
Comment (dylanbannon) [reply conf/tasks/Makefile.gke]: Yeah, I'm in favor of failing, too. I'll try removing the `||` and the `2>` and see if the output to the terminal is egregious.

Cleanup is a great idea and I've opened #274 to keep track of it.

File: .travis.yml (hunk 1 above)
Comment (dylanbannon) [reply .travis.yml]: Good idea. Done.

**Comment by dylanbannon** [comment]
@willgraf 

Alright, the testing and production workflows have been combined as much as possible - deploying a cluster in testing is almost identical to deploying one in production.

Also, these (integration) tests only run when pull-requests are open now.

File: Makefile [hunk 4]
```diff
@@ -32,3 +33,60 @@ install:
...
+	# Some debug info
+	echo "TEST"
+	printenv
+	echo $(CLOUDSDK_CORE_PROJECT) && echo $(HOME)
+	pwd
+	ls
+	make init
+	# Installations of binaries
+	## helmfile
+	@wget https://github.com/roboll/helmfile/releases/download/v0.82.0/helmfile_linux_amd64
```
Comment (willgraf) [inline Makefile]: Let's clean up this file with fewer debugging echos.

Also, I think we can move some of the "set up" functionality, like installing helm, to the `before_install` portion of travis.

Comment (dylanbannon) [reply Makefile]: I agree on both points. Done.

**Comment by dylanbannon** [comment]
@willgraf 

> Let's both do some research and understand when the best time to run the expensive tests could be. PRs maybe? merges only maybe? target branches maybe? I'm not sure.

I can't find anything definitive on this, even though it seems like it's probably a common concern. My first thought is that running whenever a PR is open is probably best, even if not ideal. BUT, there is a large problem with this...

> Another issue is the Fail-able ELK tests. I would prefer to not even test the ELK deployment than risk "successful" tests that silently fail and strand clusters. Maybe we should prune the ELK stack entirely.

Whenever we let these ELK clusters fail, we strand resources, which can cost a lot of money. I think we need to keep the ELK stack, because it was critical for us back when we were optimizing our `redis-consumer`s and I'm afraid that it may be just as necessary for people writing their custom `redis-consumer` workflows. If we can agree to keep it, then I think we just need to find a way to prevent it from costing us money.
I'm going to suggest that the easiest (it's all relative...) resolution here is to modify our deployment and destruction Makefile targets to automatically cleanup any dangling resources when things fail.

> Is there any way we can assert a cluster IS up? or a cluster IS destroyed? Can we make any assertions about the cluster state, like the helm charts that were successfully installed?

Since we're running integration tests here, it seems like the most "integrated" thing to do would be to run some workflow or series of workflows to in the running cluster before destroying it. What do you think about that?

Last point: while we're here, we might as well add unit tests. My thinking on unit tests is that most of the kiosk's unit tests should probably reside in the sub-repos (`kiosk-redis-consumer`, etc.). The only thing I can think of that should be in this repo is helmfile and helm chart linting. (Is there anything else I'm missing?)

**Comment by willgraf** [comment]
> > Let's both do some research and understand when the best time to run the expensive tests could be. PRs maybe? merges only maybe? target branches maybe? I'm not sure.
> 
> I can't find anything definitive on this, even though it seems like it's probably a common concern. My first thought is that running whenever a PR is open is probably best, even if not ideal. BUT, there is a large problem with this...

Maybe we can limit it to PR's targetting the master branch? That way we can limit some development PRs if they exist.

> > Another issue is the Fail-able ELK tests. I would prefer to not even test the ELK deployment than risk "successful" tests that silently fail and strand clusters. Maybe we should prune the ELK stack entirely.
> 
> Whenever we let these ELK clusters fail, we strand resources, which can cost a lot of money. I think we need to keep the ELK stack, because it was critical for us back when we were optimizing our `redis-consumer`s and I'm afraid that it may be just as necessary for people writing their custom `redis-consumer` workflows. If we can agree to keep it, then I think we just need to find a way to prevent it from costing us money.
> I'm going to suggest that the easiest (it's all relative...) resolution here is to modify our deployment and destruction Makefile targets to automatically cleanup any dangling resources when things fail.

I think that cleaning up all dangling resources is nearly the best-case outcome, which does not seem "easy" to me. If we temporarily disable the ELK stack (and tests), then I think we can have a separate PR for the dangling resources.

> > Is there any way we can assert a cluster IS up? or a cluster IS destroyed? Can we make any assertions about the cluster state, like the helm charts that were successfully installed?
> 
> Since we're running integration tests here, it seems like the most "integrated" thing to do would be to run some workflow or series of workflows to in the running cluster before destroying it. What do you think about that?

Ehh...that does seem like the most robust test, but I'm hesitant to want to bring up a GPU. Let's start off with some light-weight assertions about the state of the cluster, what helm charts are up, what HPAs are up, stuff like that.

Maybe future work can run through an image through the GPU, but I see this getting complicated as it is dependent on the `tensorflow-serving` model and version of `redis-consumers`, and the input image data.

> Last point: while we're here, we might as well add unit tests. My thinking on unit tests is that most of the kiosk's unit tests should probably reside in the sub-repos (`kiosk-redis-consumer`, etc.). The only thing I can think of that should be in this repo is helmfile and helm chart linting. (Is there anything else I'm missing?)

A few ideas for some unit tests:

- lint custom `helm` charts.
- lint all `helmfiles` - somehow include checking if docker tags/images exist?
- `menu.sh`
  - IDK how to test this, if at all possible. maybe pipe some commands to the menu to assert menu comes up and exits?
- some kind of `kubectl` linting for potentially `gomplated` yaml files
  - `conf/patches/redis-exporter-script.yaml`
  - `conf/patches/hpa.yaml`
  - `conf/patches/gpu-nodes.yaml`
  - `conf/addons/` maybe? I'm not sure we use that stuff

Finally, we have a LOT of logic in `conf/tasks`, as well as a few scripts in `con/patches`. I'm not sure how to test them.  I didn't find any sort of `gcloud --dry-run` option.  Maybe we can do some clever stuff with [PHONEY targets](https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html) to skip some make commands during create_all?

**Comment by dylanbannon** [comment]
> Maybe we can limit it to PR's targetting the master branch? That way we can limit some development PRs if they exist.

After doing some research and then talking with @willgraf and @MekWarrior, it looks like the best balance between code safety and fiscal safety is this:
1. We will run unit tests on every commit.
2. We will run integration tests only on PRs open into `master` and `stable` (see #280) AND only if a commit contains some special keyword phrase that we've yet to decide on. (Using this functionality: https://docs.travis-ci.com/user/conditions-v1.)
3. We will add a checkbox to the Pull Request template for the author to assert that he or she has run integration tests and to record the commit (using either the commit message or hash slug) that the tests were run on.

This *does open us up to potentially merging code into `master` and `stable` that hasn't been integration tested*, but only if both the PR author and reviewer forget to run integration tests. Given the desire to mitigate the cost and CI/CD pipeline clogging of the integration tests, and given the limitations of TravisCI's build options (why can't we just build after clicking the merge button but before the merge is actually allowed to go through?), this seems like the most sensible option.

**Comment by dylanbannon** [comment]
After futhering rumination and discussion, @willgraf, @MekWarrior, and I agreed that it'd make more sense if we modified the build policy above a little. Now,

1. Unit tests run on every commit.
2. We incorporate #280, so that we only merge into `master` from `stable`.
3. As a results of point 2, we now should be able to run tests on PRs into `master`, since they'll only be coming from `stable`, which itself shouldn't ever host much development. So, we'll trigger integration tests automatically on PRs into `master`.
4. We still modify the PR template to require integration testing (triggered by a yet-to-be-chosen keyword in a commit message) before merging into `stable`, but now it's not a huge problem if the author and reviewer of a PR into `stable` both forget to run integration tests.

This modified workflow removes the possibility of merging into `master` without running integration tests and also minimizes the number of times we run integration tests as best we feasibly can right now.

**Comment by dylanbannon** [comment]
With the latest commit, we have a solution that runs integration tests on every merge into master but can also be triggered on any commit by inserting the string `[build-integration-tests]` in the commit message. (We all wanted to choose a cool emoji as the trigger, but it seemed to me like inputting an emoji might be difficult in some cases, so I went with a boring, old text string.)

**Comment by dylanbannon** [comment]
The ELK stack test is now disabled unless a commit message contains `[test-elk]`.

To re-enable the ELK stack test, we need to resolve #274 first.

**Comment by dylanbannon** [comment]
Alright, I'm going to *say* that I've addressed all the points in the last review:

- Integration tests only run when merging into master, or when the author of a commit explicitly requests it. (N.B.: For this to be fully finished, we need to implement the changes in #280 after merging this branch.)
- ELK stack tests never run, unless a commit author explicitly requests it.
- I've opened #281 for discussion of the best workflow(s) to run on an integration test cluster.
- We have #272 open as a reminder about unit tests.

I've punted a few things to other issues, mostly just to get this PR to a merge-able state, but I think it has accomplished its main goal and I'm confident we're tracking all of the dangling issues we identified in the process.

File: .github/PULL_REQUEST.md [hunk 5]
```diff
@@ -3,3 +3,6 @@
 
 ## Why
 * Provide the justifications for the changes.
+
+## Integration Tests
+* Before requesting review, please run integration tests on your branch by making a new commit and including `[build-integration-tests]` in the commit message. Our TravisCI configuration will recognize this string and trigger integration tests. It's pretty cool. Afterwards, please indicate the relevant commit below, using either the commit message or truncated commit hash. (Note that triggering integration tests manually isn't necessary if you're merging from `stable` into `master`.)
```
Comment (willgraf) [inline .github/PULL_REQUEST.md]: It seems a bit wordy compared to the other text in the template. 

## Integration Tests
* Run integration testing by including `[build-integration-tests] in the commit message.
* Integration tested commit hash: `(         )`

Comment (dylanbannon) [reply .github/PULL_REQUEST.md]: Implemented!

**Comment by willgraf** [comment]
I was about to approve the changes to this PR, but there's no green checkmark coming up!

Maybe we can build (but not run) on all branches, but only run the integration tests on stable/master? We could mock up the unit test part here, but just have it echo "no unit tests yet" or something as a placeholder. That way it is easier to add units next issue.

Additionally, maybe we can also update the name of the stage to `integration` or `integration-tests` or something for more clarity.

**Comment by dylanbannon** [comment]
WHOO!
